### PR TITLE
Add additional iMX RT support

### LINKED ISSUE #2477: mimxrt10xx: Support the rest of the EVKs
Need to add the SDKs, linker files and board definitions for those.

Boards:
- [ ] mimxrt1015_evk
- [x] mimxrt1020_evk
- [ ] mimxrt1050_evk
- [x] mimxrt1060_evk
- [ ] mimxrt1064_evk

### LINKED ISSUE #3521: iMX RT 1020 EVK doesn't work
The 6.0.0-beta.2 release seems to crash during start up. No USB is available.

### PR DESCRIPTION
**Pull request by tannewt**
This adds a script to generate the peripherals files (except clock).

It adds support for the 1015, 1020, 1040, and 1050 EVKs.

Some work was started on 1176 but it isn't working. So, the board def is in a separate branch.

Fixes #3521. Fixes #2477.

### CODE REVIEW

File: ports/mimxrt10xx/boards/imxrt1015_evk/mpconfigboard.mk [hunk 1]
```diff
@@ -0,0 +1,8 @@
+USB_VID = 0x239A
+USB_PID = 0x8078
```
Comment (dhalbert) [inline ports/mimxrt10xx/boards/imxrt1015_evk/mpconfigboard.mk]: Assign a new PID.

File: ports/mimxrt10xx/boards/imxrt1015_evk/pins.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    { MP_OBJ_NEW_QSTR(MP_QSTR_USER_SW), MP_ROM_PTR(&pin_GPIO_EMC_09) },
+
+    // Audio Interface
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_INT), MP_ROM_PTR(&pin_GPIO_EMC_08) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_SYNC), MP_ROM_PTR(&pin_GPIO_EMC_27) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_BCLK), MP_ROM_PTR(&pin_GPIO_EMC_26) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RXD), MP_ROM_PTR(&pin_GPIO_EMC_21) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TXD), MP_ROM_PTR(&pin_GPIO_EMC_25) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_MCLK), MP_ROM_PTR(&pin_GPIO_EMC_20) },
```
Comment (dhalbert) [inline ports/mimxrt10xx/boards/imxrt1015_evk/pins.c:54]: I called this `I2S` on the Prop-Maker, and chose these pin names, to match the arg names to `I2SOut()`,
specifically:
```c
    { MP_ROM_QSTR(MP_QSTR_I2S_DATA), MP_ROM_PTR(&pin_GPIO16) },
    { MP_ROM_QSTR(MP_QSTR_I2S_BIT_CLOCK), MP_ROM_PTR(&pin_GPIO17) },
    { MP_ROM_QSTR(MP_QSTR_I2S_WORD_SELECT), MP_ROM_PTR(&pin_GPIO18) },
```
(I didn't need all the pins above.)

File: ports/mimxrt10xx/boards/imxrt1040_evk/mpconfigboard.mk [hunk 3]
```diff
@@ -0,0 +1,8 @@
+USB_VID = 0x239A
+USB_PID = 0x8084
```
Comment (dhalbert) [inline ports/mimxrt10xx/boards/imxrt1040_evk/mpconfigboard.mk]: Assign a new PID. This is the 1060 EVK PID.

File: ports/mimxrt10xx/boards/imxrt1040_evk/pins.c [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    // Audio Interface
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_INT), MP_ROM_PTR(&pin_GPIO_SD_B1_03) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_MCLK), MP_ROM_PTR(&pin_GPIO_B0_13) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RX_SYNC), MP_ROM_PTR(&pin_GPIO_B0_14) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RX_BCLK), MP_ROM_PTR(&pin_GPIO_B0_15) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RXD), MP_ROM_PTR(&pin_GPIO_AD_B1_00) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TXD), MP_ROM_PTR(&pin_GPIO_AD_B1_01) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TX_BCLK), MP_ROM_PTR(&pin_GPIO_AD_B1_02) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TX_SYNC), MP_ROM_PTR(&pin_GPIO_AD_B1_03) },
```
Comment (dhalbert) [inline ports/mimxrt10xx/boards/imxrt1040_evk/pins.c:87]: same comment on I2S pins as above

File: ports/mimxrt10xx/boards/imxrt1050_evkb/mpconfigboard.mk [hunk 5]
```diff
@@ -0,0 +1,10 @@
+USB_VID = 0x239A
+USB_PID = 0x8084
```
Comment (dhalbert) [inline ports/mimxrt10xx/boards/imxrt1050_evkb/mpconfigboard.mk]: Need a new PID. This is the 1060 PID.

File: ports/mimxrt10xx/boards/imxrt1050_evkb/pins.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    // Audio Interface
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_INT), MP_ROM_PTR(&pin_GPIO_AD_B1_08) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_MCLK), MP_ROM_PTR(&pin_GPIO_AD_B1_09) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RX_SYNC), MP_ROM_PTR(&pin_GPIO_AD_B1_10) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RX_BCLK), MP_ROM_PTR(&pin_GPIO_AD_B1_11) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_RXD), MP_ROM_PTR(&pin_GPIO_AD_B1_12) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TXD), MP_ROM_PTR(&pin_GPIO_AD_B1_13) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TX_BCLK), MP_ROM_PTR(&pin_GPIO_AD_B1_14) },
+    { MP_ROM_QSTR(MP_QSTR_AUDIO_TX_SYNC), MP_ROM_PTR(&pin_GPIO_AD_B1_15) },
```
Comment (dhalbert) [inline ports/mimxrt10xx/boards/imxrt1050_evkb/pins.c:101]: `I2S`

File: ports/mimxrt10xx/common-hal/microcontroller/Processor.c [hunk 7]
```diff
@@ -50,6 +53,9 @@ float common_hal_mcu_processor_get_temperature(void) {
     OCOTP_Deinit(OCOTP);
 
     return temp;
+    #else
+    return 0;
```
Comment (dhalbert) [inline ports/mimxrt10xx/common-hal/microcontroller/Processor.c:57]: I'd prefer to raise `NotImplementedErrror` here. I looked at the other ports. Espressif raises if temp is not available. STM returns `false`  (? - that is, zero), a few ports return `NaN`, which I think is kind of a bad idea. We should fix all those to raise.

Comment (dhalbert) [reply ports/mimxrt10xx/common-hal/microcontroller/Processor.c:57]: same for `get_voltage` below

File: lib/tinyusb
Comment (jepler) [inline lib/tinyusb:1]: is this change intentional? (bringing in 11xx support?)

**Review by dhalbert — DISMISSED** [review]
Glad you got these working. A few minor things.

I think it would be better to call these boards `nxp_mimx...` instead of just `mimx...`. I know the first one was not named that. I would really like to see the manufacturer included in the board name. I know we haven't been consistent, and it's going to be a job to rename them all. So if you don't want to do it now and we do it in a big bang later, that's OK.

File: ports/mimxrt10xx/boards/metro_m7_1011/flash_config.c [hunk 8]
```diff
@@ -14,14 +14,14 @@ __attribute__((section(".boot_hdr.conf")))
...
-    .ipcmdSerialClkFreq = kFLEXSPISerialClk_30MHz,
+    .ipcmdSerialClkFreq = kFLEXSPISerialClk_133MHz,
     .blockSize = 0x00010000,
     .isUniformBlockSize = false,
     .memConfig =
     {
         .tag = FLEXSPI_CFG_BLK_TAG,
         .version = FLEXSPI_CFG_BLK_VERSION,
-        .readSampleClkSrc = kFLEXSPIReadSampleClk_LoopbackFromDqsPad,
+        .readSampleClkSrc = kFLEXSPIReadSampleClk_LoopbackFromSckPad,
```
Comment (jepler) [inline ports/mimxrt10xx/boards/metro_m7_1011/flash_config.c:24]: what does this change _mean_? It's nice, if it enables the flash to be clocked higher.

File: ports/mimxrt10xx/tools/gen_peripherals_data.py
Comment (jepler) [inline ports/mimxrt10xx/tools/gen_peripherals_data.py:1]: is it feasible to run this script at build time and remove the committed files?   if the file is auto-generated and **never** intended to be hand-edited it would be nice to have a comment that says so, too.

**Review by jepler — CHANGES_REQUESTED** [review]
Thanks -- I have some questions though.

File: lib/tinyusb
Comment (tannewt) [reply lib/tinyusb:1]: It is because it brings in 1040 support. (It names its USB weird.)

File: ports/mimxrt10xx/tools/gen_peripherals_data.py
Comment (tannewt) [reply ports/mimxrt10xx/tools/gen_peripherals_data.py:1]: I don't think so because it depends on files from their configurator tool that I don't think can be redistributed. I'll add a comment. I was going for minimal diff.

File: ports/mimxrt10xx/boards/metro_m7_1011/flash_config.c (hunk 8 above)
Comment (tannewt) [reply ports/mimxrt10xx/boards/metro_m7_1011/flash_config.c:24]: It means we can use/reset the DQS pad. (Otherwise flash stops working when you reset DQS even when not connected externally.)

File: ports/mimxrt10xx/boards/imxrt1015_evk/pins.c (hunk 2 above)
Comment (tannewt) [reply ports/mimxrt10xx/boards/imxrt1015_evk/pins.c:54]: I think I want to leave these as-is. We can add standard names later when we test this. The extra pins are used for I2S *input* and I don't want to use the names above because they don't include direction info. (The 1040 and 1050 have separate bit clock and word select lines per direction too.) The peripheral can do other protocols besides I2S so I think the primary name should be generic.

File: ports/mimxrt10xx/common-hal/microcontroller/Processor.c (hunk 7 above)
Comment (tannewt) [reply ports/mimxrt10xx/common-hal/microcontroller/Processor.c:57]: The shared-bindings are documented that they return None when not available. They convert NAN to None so NAN matches the docs.

File: ports/mimxrt10xx/boards/imxrt1015_evk/pins.c (hunk 2 above)
Comment (dhalbert) [reply ports/mimxrt10xx/boards/imxrt1015_evk/pins.c:54]: > The peripheral can do other protocols besides I2S so I think the primary name should be generic.

I did not know, sounds good.

**Review by jepler — APPROVED** [review]
thanks for addressing my concern!
